Subtarget for the CV32E40P and respective tests

Cecil-W opened the pull request:
A simple sub-target for the cv32e40p. When benchmarking coremark, using core-v-verif, TuneNoDefaultUnroll provided slight improvements.

Cecil-W committed:
Subtarget for the CV32E40P and respective tests

lewis-revill commented:
Perhaps we could come up with a test for the changes in the scheduling model, I think there are some already-existing tests in the backend somewhere which will most likely be machine IR -> machine IR tests. For now we'll keep the model as it is but in future we may want a no-floating-point variant

simonpcook commented on llvm/include/llvm/Support/RISCVTargetParser.def:
```diff
@@ -19,6 +19,7 @@ PROC(SIFIVE_S54, {"sifive-s54"}, FK_64BIT, {"rv64gc"})
 PROC(SIFIVE_S76, {"sifive-s76"}, FK_64BIT, {"rv64gc"})
 PROC(SIFIVE_U54, {"sifive-u54"}, FK_64BIT, {"rv64gc"})
 PROC(SIFIVE_U74, {"sifive-u74"}, FK_64BIT, {"rv64gc"})
+PROC(CV32E40P, {"CV32E40P"}, FK_NONE, {"rv32imfc"})
```
Two minor tweaks here, I'm not 100% sure on the exact answer for these, but interested in comments.

1. This list looks like it was sorted before adding this, so should cv32e40p be moved up the list. (Subpoint, should generic be top of the list regardless?)
2. For consistency, I think perhaps cv32e40p might want to be made lowercase.

Cecil-W replied:
Regarding the order in RISCVTargetParser.def, i can move the list, but i think keeping the generic targets at the top makes sense.
Also, this file is already removed in upstream llvm since this [commit](https://github.com/llvm/llvm-project/commit/ac1ffd3caca12c254e0b8c847aa8ce8e51b6cfbf) as they moved to a tablegen based generation for the cpu definitions.

Regarding 2.: Lowercase probably makes more sense, i will update it.

Cecil-W commented:
> For now we'll keep the model as it is but in future we may want a no-floating-point variant.
 
This would be very easy to add as the model can be reused for a new processor without the "FeatureStdExtF". 
For example the sifive-e20 does it in the same way as it uses the RocketModel.

Cecil-W committed:
changed the processor name to lowercase for consistency with others

PhilippvK commented on llvm/lib/Target/RISCV/RISCVSchedcv32e40p.td:
```diff
@@ -0,0 +1,222 @@
+def : WriteRes<WriteLDWU, [cv32e40pUnitLSU]>;
+def : WriteRes<WriteLDB, [cv32e40pUnitLSU]>; // lbu
+
+// Jump instructions, assuming jumps are word alligned
```
Typo: aligned

PhilippvK commented on llvm/lib/Target/RISCV/RISCVSchedcv32e40p.td:
```diff
@@ -0,0 +1,222 @@
+  def : WriteRes<WriteFSqrt32, [cv32e40pUnitFPU]>;
+}
+
+
```
Please re-format to remove whitespace etc.

Cecil-W committed:
Reformatting and fixed typo

ChunyuLiao commented:
I would like to ask a question:
What is the purpose of adding scheduling model? I didn't find specific documentation for the cycles information.

Cecil-W commented:
I created this model during my research internship and because TuneNoDefaultUnroll improved performance  in coremark i thought it may be worth to contribute. 
The cycle information was taken from the [user manual](https://docs.openhwgroup.org/_/downloads/cv32e40p-user-manual/en/latest/pdf/) chapter 12.2 (at the time of writing this comment).

PhilippvK commented:
Not ready for merge yet. @Cecil-W will submit a refined version soon. Also, I would like to come up with a representative unit test.

Cecil-W committed:
modeled data hazards regarding loads and jalr

Cecil-W committed:
Modeled the latency of mulh* correctly

Cecil-W committed:
Enabled the PostRAScheduler
and added missing ResourceCycles

Cecil-W committed:
Changed the model to a single ProcResource.
This is similar to other single-issue architectures like the cortex M4.
It performes the same and makes the model simpler
Fixed a few Typos

PhilippvK commented on llvm/lib/Target/RISCV/RISCVSchedcv32e40p.td:
```diff
@@ -0,0 +1,219 @@
+//===----------------------------------------------------------------------===//
+// ReadAdvance definitions
+def : ReadAdvance<ReadJmp, 0>;
+// There is a 1-cycle data hazard if jalr depends on a immediatly preceding instrucion
```
typo: immediately

Cecil-W replied:
Fixed

PhilippvK commented on llvm/lib/Target/RISCV/RISCVSchedcv32e40p.td:
```diff
@@ -0,0 +1,219 @@
+// ===---------------------------------------------------------------------===//
```
Apache License header missing here

PhilippvK replied:
Also, please refer to https://cv32e40p.readthedocs.io/en/latest/pipeline.html

Cecil-W replied:
I added the missing Apache License header.

Could you explain, to what i should refer in the docs? As far as i can tell, the latency in the schedule should be correct.

PhilippvK replied:
My intention was just putting a reference to the spec somewhere in the code in cause someone wants to look up the latencies by themselves.

Cecil-W committed:
Added Apache License header

Cecil-W committed:
Fixed a typo

Cecil-W committed:
Added a link to the pipeline documentation
